tests/testthat/helper.R

```diff
@@ -1,3 +1,75 @@
+#' Does code throw an rextendr_error?
+#'
+#' `expect_rextendr_error()` expects an error of class `rextendr_error`, as
+#' thrown by `ui_throw()`.
+#'
+#' @param ... arguments passed to [testthat::expect_error()]
 expect_rextendr_error <- function(...) {
   expect_error(..., class = "rextendr_error")
 }
+
+#' Create a local package
+#'
+#' `local_package()` creates a self-cleaning test package via usethis and withr.
+#' It also sets the local working directory and usethis project to the temporary
+#' package. These settings are reverted and the package removed via
+#' `withr::defer()`. This clean-up happens at the end of the local scope,
+#' usually the end of a `testthat()` call.
+#'
+#' @param nm The name of the temporary package
+#' @param envir An environment where `withr::defer()`'s exit handler is
+#'   attached, usually the `parent.frame()` to exist locally
+#'
+#' @return A path to the root package directory
+local_package <- function(nm, envir = parent.frame()) {
+  local_temp_dir(envir = envir)
+  dir <- usethis::create_package(nm)
+  setwd(dir)
+  local_proj_set(envir = envir)
+
+  invisible(dir)
+}
+
+#' Create a local temporary directory
+#'
+#' `local_temp_dir()` creates a local temporary directory and sets the created
+#' directory as the working directory. These are then cleaned up with
+#' `withr::defer()` at the end of the scope, usually the end of the `testthat()`
+#' scope.
+#'
+#' @param envir An environment where `withr::defer()`'s exit handler is
+#'   attached, usually the `parent.frame()` to exist locally
+#'
+#' @return A path to the temporary directory
+local_temp_dir <- function(envir = parent.frame()) {
+  current_wd <- getwd()
+  path <- tempfile()
+  dir.create(path)
```

Comment:
Is there a reason `tempfile()` is used instead of `tempdir()`?

Reply:
`tempdir()` is a function that returns the parent directory of all temporary files, which contains various temporary files. So, if we want a clean temporary directory, `dir.create(tempfile())` is what we want.

``` r
tempdir()
#> [1] "/tmp/RtmpbzxQ5a"
tempfile()
#> [1] "/tmp/RtmpbzxQ5a/file185566cc1191"
```

<sup>Created on 2021-04-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

Reply:
I see, thanks @yutannihilation !

Reply:
Yeah, those names are a bit misleading IMO. I have learned the hard way that `tempdir()` is actually a very important directory, and if you mess with it too much (or delete it 😇) you get all sorts of bizarre errors in R